### PR TITLE
FSPT-664 Export Submissions as CSV

### PR DESCRIPTION
**Pull request by srh-sloan**
## 🎫 Ticket
https://mhclgdigital.atlassian.net/browse/FSPT-664

## 📝 Description
Allows platform admin users to export all submissions for a collection as a csv file. Does not yet include an e2e tests for this, will be included in a separate PR.

Includes some additions to the factories used in creating a collection to create completed submissions, either with random or known data.

## 📸 Show the thing (screenshots, gifs)
https://github.com/user-attachments/assets/49434861-3297-411d-9076-097f34525517

## 🧪 Testing
- Create a test submission
- Click 'export as CSV'
- Open the downloaded file - it should display correctly in a text editor (eg. vs code) or a spreadsheet tool (eg. Numbers)

### Review Readiness
- [x] PR title and description provide sufficient context for reviewers
- [ ] Commits are logical, self-contained, and have clear descriptions

### Performance and security
- [x] No N+1 query problems introduced
- [n/a] Any new DB queries include appropriate where clauses based on the user's permissions

### Testing
- [x] I have tested this change and it meets the acceptance criteria for the ticket
- [ ] I need the reviewer(s) to pull and run this change locally
- [ ] New (non-developer) functionality has appropriate unit and integration tests
- [TODO] End-to-end tests have been updated (if applicable)
- [x] Edge cases and error conditions are tested

### CODE REVIEW

File: app/common/helpers/collections.py [hunk 1]
```diff
@@ -323,6 +334,73 @@ def get_previous_question(self, current_question_id: UUID) -> Optional["Question
...
+        for submission in [value for key, value in self.submission_helpers.items()]:
+            submission_csv_data = {
+                "Submission reference": submission.reference,
+                "Created by": submission.created_by_email,
+                "Created time UTC": submission.created_at_utc.isoformat(),
+            }
+            visible_questions = submission.all_visible_questions
+            for question_id, header_string in question_headers.items():
+                if question_id not in visible_questions.keys():
+                    submission_csv_data[header_string] = "Not required"
```
Comment (samuelhwilliams) [inline app/common/helpers/collections.py:392]:
```suggestion
                    submission_csv_data[header_string] = "NOT_ASKED"
```

Maybe saying 'not asked' instead of 'not required' will distinguish between questions that are mandatory which weren't shown, and questions which are optional and not answered?

Currently all questions are mandatory but not forever, we think.

Possibly also UPPERCASE_UNDERSCORES a bit more indicative of a system value rather than a user entered string?

File: tests/models.py [hunk 2]
```diff
@@ -136,56 +137,131 @@ class Meta:
...
+            _create_submission(SubmissionModeEnum.LIVE, complete_question_2=False)
+
+    @factory.post_generation  # type: ignore
+    def create_completed_submissions_each_question_type(  # type: ignore
         obj: Collection,
         create,
         extracted,
         test: int = 0,
         live: int = 0,
+        use_random_data: bool = True,
```
Comment (samuelhwilliams) [inline tests/models.py:207]: I think all of the `use_random_data` do nothing now and can be dropped?

File: app/common/helpers/collections.py [hunk 3]
```diff
@@ -323,6 +334,73 @@ def get_previous_question(self, current_question_id: UUID) -> Optional["Question
         raise ValueError(f"Could not find a question with id={current_question_id} in collection={self.collection}")
 
 
+class CollectionHelper:
```
Comment (samuelhwilliams) [inline app/common/helpers/collections.py:337]: Adding this new higher-level helper feels good 👍

File: app/common/helpers/collections.py [hunk 4]
```diff
@@ -323,6 +334,73 @@ def get_previous_question(self, current_question_id: UUID) -> Optional["Question
...
+        Returns a list of all questions that are part of the collection, across all sections and forms.
+        """
+        return [
+            question
+            for section in sorted(self.collection.sections, key=lambda s: s.order)
+            for form in sorted(section.forms, key=lambda f: f.order)
+            for question in sorted(form.questions, key=lambda q: q.order)
+        ]
+
+    def generate_csv_content_for_all_submissions(self) -> str:
```
Comment (samuelhwilliams) [inline app/common/helpers/collections.py:372]: This has a risk of becoming quite a heavy operation when we have lots of submissions and collections are complicated - any thought on how we monitor it for performance and optimise it in the future (eg generators/streaming/background tasks)? I think definitely one to think about a bit carefully now with the new PR template around performance/n+1 queries etc.

File: app/developers/deliver_routes.py [hunk 5]
```diff
@@ -952,11 +951,38 @@ def list_submissions_for_collection(collection_id: UUID, submission_mode: Submis
...
     )
 
 
+@developers_deliver_blueprint.route(
+    "/collections/<uuid:collection_id>/submissions/<submission_mode:submission_mode>/export",
+    methods=["GET"],
+)
+@is_platform_admin
+def export_submissions_for_collection(collection_id: UUID, submission_mode: SubmissionModeEnum) -> ResponseReturnValue:
+    export_format = request.args.get("format", "")
```
Comment (samuelhwilliams) [inline app/developers/deliver_routes.py:966]:
```suggestion
@developers_deliver_blueprint.route(
    "/collections/<uuid:collection_id>/submissions/<submission_mode:submission_mode>/export/<export_format>",
    methods=["GET"],
)
@is_platform_admin
def export_submissions_for_collection(collection_id: UUID, submission_mode: SubmissionModeEnum, export_format: str) -> ResponseReturnValue:
```

optional nit: could embed the export type in the URL so it's a bit more discoverable than a random query arg?

File: app/common/helpers/collections.py (hunk 4 above)
Comment (MarcUsher) [reply app/common/helpers/collections.py:372]: Yeah this was the big issue on Find wasn't it, though I guess that was generating an excel file rather than csv, but not beyond the realms of possibility that we end up with monitoring collections that would be better as excel files with tabs rather than single CSVs with cross-referenced tabs etc.

Comment (srh-sloan) [reply app/common/helpers/collections.py:373]: To be address in a separate PR

**Review by MarcUsher — APPROVED** [review]
This is really great Sarah 👌
